File: Practice/AllDomains/Tutorials/CrackingTheCodingInterview/QueuesATaleofTwoStacks.cs
/*	author:@shivkrthakur  */
using System;
using System.Collections.Generic;
using System.IO;
class Solution {
    static void Main(String[] args) {
        /* Enter your code here. Read input from STDIN. Print output to STDOUT. Your class should be named Solution */
        int n = Convert.ToInt32(Console.ReadLine().Trim());
        Stack<int> stack1 = new Stack<int>();
        Stack<int> stack2 = new Stack<int>();
        
        for(int x = 0; x < n; x++)
        {
            int [] arr = Array.ConvertAll(Console.ReadLine().Trim().Split(),Int32.Parse);
            int queryType = arr[0];
            if(queryType == 1)//ENQUEUE
            {
                int pushVal = arr[1];
                stack1.Push(pushVal);
                //Console.WriteLine("Query Type is 1, Pushed: " + pushVal + " Stack1 Count is : " + stack1.Count);
            }
            else if(queryType == 2)//DEQUEUE
            {
                int popVal = 0;
                if(stack2.Count == 0)
                {
                    while(stack1.Count > 0)
                    {
                        int val = stack1.Pop();
                        stack2.Push(val);
                    }
                }
                
                if(stack2.Count > 0) popVal = stack2.Pop();
                //Console.WriteLine("Query Type is 2, Poped: " + popVal + " Stack2 Count is : " + stack2.Count);
            }
            else if(queryType == 3)//PEEK
            {
                int peekVal = 0;
                if(stack2.Count == 0)
                {
                    while(stack1.Count > 0)
                    {
                        int val = stack1.Pop();
                        stack2.Push(val);
                    }
                }
                
                if(stack2.Count > 0) peekVal = stack2.Peek();  
                //Console.WriteLine("Query Type is 3, Peeked: " + peekVal + " Stack2 Count is : " + stack2.Count);
                Console.WriteLine(peekVal);
            }
        }
    }
}

File: Practice/AllDomains/Tutorials/CrackingTheCodingInterview/QueuesATaleofTwoStacks.js
/*	author:@shivkrthakur  */
var inputArr = [];
var index = 0;

function readLine(){
    return inputArr[index++];
}

function processData(input) {
    //Enter your code here
    inputArr = input.split('\n');
    var q = parseInt(readLine());
    var stack1 = [], stack2 = [];
    for(var i = 0; i < q; i++){
        var ip = readLine().split(' ').map(Number);
        //console.log(ip);
        if(ip[0] == 1){
            stack1.push(ip[1]);
        }
        if(ip[0] == 2){
            while(stack1.length > 0){
                stack2.push(stack1.pop())
            }
            stack2.pop();
            while(stack2.length > 0){
                stack1.push(stack2.pop())
            }
        }
        if(ip[0] == 3){
            console.log(stack1[0]);
        }
    }
}

process.stdin.resume();
process.stdin.setEncoding("ascii");
_input = "";
process.stdin.on("data", function (input) {
    _input += input;
});

process.stdin.on("end", function () {
   processData(_input);
});
